### ISSUE TITLE
[BUG] Checker allows to unsoundly override a @SafeEffect method as @UIEffect if the method is inherited indirectly

### ISSUE THREAD

**Issue by PawelLipski** [typetools/checker-framework#3287]
**Checker version**: 3.4.0

**Affected checkers**: `GuiEffectChecker`, but possibly a similar problems occurs in other checkers where method annotations are taken into account.

**The problem**:

Given the 2 following interfaces (actually, most likely those can be classes as well):

```java
public interface IBaseInterface {
  void run();
}
```

```java
interface IDerivedInterface extends IBaseInterface { }
```

and a class implementing/overriding the `@SafeEffect` method as `@UIEffect`:

```java
public class ImplementingClass implements IDerivedInterface {
  @Override
  @UIEffect
  public void run() {}
}
```

we should get:

`error: [[all, guieffect]:override.effect.invalid] A method override may only be @UI if it overrides an @UI method.`

but instead the compilation completes successfully.

**Sample repo where the problem can be reproduced**: see the following PR, just run `./gradlew compileJava`: https://github.com/PawelLipski/git-machete-intellij-plugin/pull/1

**Comment by PawelLipski** [comment]
Also, out of curiosity: why does the message say `A method override may only be @UI if it overrides an @UI method` instead of `A method override may only be @UIEffect if it overrides an @UIEffect method`? `@UI` annotation doesn't even target methods.

**Comment by csgordon** [comment]
That message text may date back to before ```@UIEffect``` was named as it is now (a long time ago), and should probably be fixed as well.

**Comment by csgordon** [comment]
The bug only arises in cases where the inherited method comes from an interface; if a direct super*class* defines the method as ```@SafeEffect``` the inheritance check catches invalid overrides at ```@UIEffect```.  A method I thought reflected semantically present elements only reflects elements present directly in the textual source.  A naive fix is relatively straightforward: search all transitively-reachable parent interfaces (whether the class implements an interface or subinterface, or extends a class that implements such an interface).  I'll prototype a naive solution (brute force search) as a baseline, but then I'll implement a caching search so running time doesn't tank on more complex inheritance hierarchies.

**Comment by PawelLipski** [comment]
@csgordon any progress on this one? :)

**Comment by csgordon** [comment]
Sorry for the delay, @PawelLipski. I have a buggy prototype of a different traversal of the inheritance hierarchy, which follows what should be the correct approach, but my changes broke some functionality I thought I had stopped using and it now misses some obvious errors.  I know roughly what to do, but have had trouble finding time for it. I expect to loop back to this problem this week or next.

**Comment by csgordon** [comment]
Sorry for the slow turnaround @PawelLipski.  Commit 23c6458 on [csgordon/checker-framework](https://github.com/csgordon/checker-framework) should fix it, and has a clean CI run.  If you're willing to build yourself, you could try it out; it correctly handles the nicely reduced test case you provided plus another variation.

**Comment by PawelLipski** [comment]
Thank you @csgordon  :)
